### ISSUE TITLE
v2.3.0 gives internal nil pointer on "phase.RLock()" but 2.2.1 does work.

### ISSUE THREAD

**Issue by iangregsondev** [projectdiscovery/naabu#1059]
### Naabu version: GO Library 2.3.0 (See packages used below)

This is using the go library, and two packages being imported are

```
	"github.com/projectdiscovery/naabu/v2/pkg/result"
	"github.com/projectdiscovery/naabu/v2/pkg/runner"
```

These are all set to 2.3.0 (the latest) at the time of writing.

### Current Behavior:
I am using Nuclei and Naabu, but they have entirely separate functions. If I run a nuclei scan then a nil pointer and a stack trace is coming from "naabu", I will paste this here :-

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x102fe4e64]

goroutine 52 [running]:
github.com/projectdiscovery/naabu/v2/pkg/scan.(*Phase).Is(0x14001b86c01?, 0x2?)
        /Users/ian/.gvm/pkgsets/go1.22/global/pkg/mod/github.com/projectdiscovery/naabu/v2@v2.3.0/pkg/scan/scan.go:50 +0x24
github.com/projectdiscovery/naabu/v2/pkg/scan.TransportReadWorker.func1({{{0x14008af32e2, 0x2c, 0x2c}, {0x14008af32e2, 0x0, 0x0}}, 0x50, 0xcd2e, 0xc152c4e9, 0x9d744071, ...}, ...)
        /Users/ian/.gvm/pkgsets/go1.22/global/pkg/mod/github.com/projectdiscovery/naabu/v2@v2.3.0/pkg/scan/scan_unix.go:528 +0xbc
github.com/projectdiscovery/naabu/v2/pkg/scan.TransportReadWorker.func3(0x14000b986c0)
        /Users/ian/.gvm/pkgsets/go1.22/global/pkg/mod/github.com/projectdiscovery/naabu/v2@v2.3.0/pkg/scan/scan_unix.go:641 +0x55c
created by github.com/projectdiscovery/naabu/v2/pkg/scan.TransportReadWorker in goroutine 35
        /Users/ian/.gvm/pkgsets/go1.22/global/pkg/mod/github.com/projectdiscovery/naabu/v2@v2.3.0/pkg/scan/scan_unix.go:599 +0x140
task: Failed to run task "module": exit status 2
```

It's worth mentioning that this is NOT running any NAABU code. Only the "imports" have been executed because they are in a "discovery" struct that I set up—but no code is being executed. I believe that some NAABU code is executed on valid imports because it uses a number of "init" functions.

I tried debugging and noticed that "phase" is nil in this struct / method in the naabu source code.

```
func (phase *Phase) Is(state State) bool {
	phase.RLock() ///// <<<< This is where the issue is. phase is nil.
	defer phase.RUnlock()

	return phase.State == state
}
```

From what I understand, nuclei does not depend on any naabu code.

Rolling naabu back to 2.2.1 fixes this issues.

The current test machine is a Mac M1.

### Expected Behavior:
There should be no internal panic on v2.3.0

**Comment by Mzack9999** [comment]
@iangregsondev Thanks for opening the issue, I've tried with a minimal example but unable to reproduce:
```go
package main

import (
	"time"

	_ "github.com/projectdiscovery/naabu/v2/pkg/result"
	_ "github.com/projectdiscovery/naabu/v2/pkg/runner"
)

func main() {
	time.Sleep(time.Hour)
}
```
I guess it might be due to the way the struct is initialized, probably invoking some function. Would it be possible to provide a minimal example to reproduce the behavior, stripping out anything sensitive?

**Comment by iangregsondev** [comment]
Yes, you need to run a nuclei scan at the same time....

Its weird, if you just import - it doesn't happen. But having the code imported and then running a nuclei scan via go, it happens.

My code is wrapped in some custom code, but I can extract it and post it here. I will try and do this tomorrow.

Thanks.

**Comment by iangregsondev** [comment]
One thing I did notice @Mzack9999 is that I wanted to make my example easy for you to test... I am using IP's.

SO I thought I would place the following..

as a target "scanme.sh"

And it did not happen :-(

I have a number of other IP only values on my LAN and it does happen here

"192.168.1.1:80"

and 

"192.168.1.50:80"

I tried each one and I get the issue but not with scanme.sh :-(

192.168.1.1:80 is just a router I have

I don't remember what 192.168.1.50:80 is :-) I think its a linux machine.

Then I had a thought, maybe I should just pass down 192.168.1.1 without the port number for the nuclei scan... and I did and it did and naabu did not panic either..

So maybe its a port thing - very confused now :-)

Should I create a working example and create  some debug/verbose log output - maybe that will help.

I am assuming that if you pass any IP with a port to nuclei then it will cause naabu to panic.

**Comment by iangregsondev** [comment]
Hi @Mzack9999 ,

As promised, here is the example.

There is a README,

There is no execution of naabu, only execution of nuclei but the panic outputs, it takes about 3 - 4 minutes to output on my system.

https://github.com/iangregsondev/naabu_panic

**Comment by iangregsondev** [comment]
In the readme, i have placed the panic info.

**Comment by Mzack9999** [comment]
@iangregsondev Thanks for providing the PoC, I'll have a look shortly

**Comment by iangregsondev** [comment]
Thats great, thank you. If I can provide anything else, let me know.

With regards to not being able to change the templates directory and having to downloading the nuclei templates..

I think I was missing this in my code

```
defaultOpts.Templates =  goflags.StringSlice{cpath}
```

But anyway, I tested it by downloading the templates manually into the home directory as per my README.

Thanks again.

**Comment by iangregsondev** [comment]
@Mzack9999 Something strange, my test target (see code) is this. 

scanme.sh

```
		ctx, networkscanner.ScanCommandData{
			ID: "1234",
			// Ranges: []string{"scanme.sh"},
			Ranges: []string{"192.168.1.1:80"},
```

It was failing and then I found it started passing :-(

Anyway, replacing this with my router IP does fail again.

Only calling this out, in case you don't see an issue, if you have some router on your LAN or something, you could try that.

Thanks.

**Comment by Mzack9999** [comment]
@iangregsondev Thanks for the very detailed PoC and additional info. I was not able to reproduce the issue, but I think I found what could have provoked it. Would it be possible to checkout https://github.com/projectdiscovery/naabu/pull/1063 within your project and see if the issue got fixed? Thanks!

```console
go get -v github.com/projectdiscovery/naabu/v2@01a3accc9553fa7103b38d7581fe3700fa3e90a9
```

**Comment by iangregsondev** [comment]
@Mzack9999 Thank you!

Just been testing it, and it seems to have fixed it!! :-)

I have run it against my IP, which always seems to fail. I tested it prior to installing the update, and it failed.

After installing the update, it no longer fails. I have ran it a number of times.

Looks like it's a good fix.

I presume this will end up being released as 2.3.1? Is there an approximate release date for this, just out of interest?

Again, thanks for everything and your quick response on this.
